mmlearn/modules/layers/transformer_block.py

```diff
@@ -0,0 +1,167 @@
+"""Transformer Block and Embedding Modules for Vision Transformers (ViT)."""
+
+import numpy as np
+import torch
+import torch.nn as nn
+from mmlearn.modules.layers.attention import Attention, DropPath
+from mmlearn.modules.layers.mlp import MLP
+from typing import Optional, List, Callable, Union
+
+class Block(nn.Module):
+    """
+    Transformer Block.
+
+    This module represents a Transformer block that includes self-attention, 
+    normalization layers, and a feedforward multi-layer perceptron (MLP) network.
+
+    Parameters
+    ----------
+    dim : int
+        The input and output dimension of the block.
+    num_heads : int
+        Number of attention heads.
+    mlp_ratio : float, optional, default=4.0
+        Ratio of hidden dimension to the input dimension in the MLP.
+    qkv_bias : bool, optional, default=False
+        If True, add a learnable bias to the query, key, value projections.
+    qk_scale : float, optional, default=None
+        Override default qk scale of head_dim ** -0.5 if set.
+    drop : float, optional, default=0.0
+        Dropout probability for the output of attention and MLP layers.
+    attn_drop : float, optional, default=0.0
+        Dropout probability for the attention scores.
+    drop_path : float, optional, default=0.0
+        Stochastic depth rate, a form of layer dropout.
+    act_layer : Callable[..., nn.Module], optional, default=nn.GELU
+        Activation layer in the MLP.
+    norm_layer : Callable[..., nn.Module], optional, default=nn.LayerNorm
+        Normalization layer.
+
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        num_heads: int,
+        mlp_ratio: float = 4.0,
+        qkv_bias: bool = False,
+        qk_scale: Optional[float] = None,
+        drop: float = 0.0,
+        attn_drop: float = 0.0,
+        drop_path: float = 0.0,
+        act_layer: Callable[..., nn.Module] = nn.GELU,
+        norm_layer: Callable[..., nn.Module] = nn.LayerNorm
+    ) -> None:
+        super().__init__()
+        self.norm1 = norm_layer(dim)
+        self.attn = Attention(
+            dim, num_heads=num_heads, qkv_bias=qkv_bias, qk_scale=qk_scale, attn_drop=attn_drop, proj_drop=drop
+        )
+        self.drop_path = DropPath(drop_path) if drop_path > 0. else nn.Identity()
+        self.norm2 = norm_layer(dim)
+        mlp_hidden_dim = int(dim * mlp_ratio)
+        self.mlp = MLP(in_dim=dim, hidden_dims=[mlp_hidden_dim], activation_layer=act_layer, dropout=drop)
+
+    def forward(self, x: torch.Tensor, return_attention: bool = False) -> Union[torch.Tensor, torch.Tensor]:
+        """Forward pass through the Transformer Block."""
+        y, attn = self.attn(self.norm1(x))
+        if return_attention:
+            return attn
+        x = x + self.drop_path(y)
+        x = x + self.drop_path(self.mlp(self.norm2(x)))
+        return x
+
+
+class PatchEmbed(nn.Module):
```

Comment:
Let's create a new module for the embedding layers, including the functions in `attention.py`?

Also, please consider moving [this class](https://github.com/VectorInstitute/mmlearn/blob/main/mmlearn/datasets/processors/tokenizers.py#L133) to the same module.

Reply:
Done!